src/hardware/disk_noise.cpp

```diff
@@ -0,0 +1,451 @@
+
+	if (hdd_noise) {
+		hdd_noise->Shutdown();
+		hdd_noise.reset();
+	}
+	if (active_devices.empty() && mix_channel) {
+		mix_channel->Enable(false);
+		MIXER_DeregisterChannel(mix_channel);
+		mix_channel.reset();
+	}
+}
+
+std::vector<float> DiskNoiseDevice::GetSample()
+{
+	float sample_l = 0.0f;
+	float sample_r = 0.0f;
+
+	// Mix in spinup and spin samples
+	if (!spin.spin_up_sample.empty() &&
+	    spin.spin_up_pos + 1 < spin.spin_up_sample.size()) {
+		sample_l += spin.spin_up_sample[spin.spin_up_pos] * DisknoiseGain;
+		sample_r += spin.spin_up_sample[spin.spin_up_pos++] * DisknoiseGain;
+	} else if (!spin.sample.empty() &&
+	           (spin.pos + 1 < spin.sample.size() || spin.loop)) {
+		sample_l += spin.sample[spin.pos] * DisknoiseGain;
+		sample_r += spin.sample[spin.pos++] * DisknoiseGain;
+
+		// Loop the spin sound if enabled. Used for
+		// persistent HDD noise Not used for floppy
+		// noise because motor should stop after r/w
+		// operations aredone
+		if (spin.pos >= spin.sample.size() && spin.loop) {
+			spin.pos = 0;
+		}
+	}
+
+	// Mix in seek sample, if it's playing
+	if (!seek.current_sample.empty() &&
+	    seek.pos + 1 < seek.current_sample.size()) {
+		sample_l += seek.current_sample[seek.pos] * DisknoiseGain;
+		sample_r += seek.current_sample[seek.pos++] * DisknoiseGain;
+	}
+
+	// If seek sample has finished playing, clear entry
+	if (!seek.current_sample.empty() && seek.pos >= seek.current_sample.size()) {
+		seek.current_sample.clear();
+	}
+
+	return {sample_l, sample_r};
+}
+
+void DiskNoiseDevice::LoadSample(const std::string& path, std::vector<float>& buffer)
+{
+	if (path.empty()) {
+		return;
+	}
+
+	constexpr auto SampleExtension = ".flac";
+
+	const auto candidate_paths = {std_fs::path(path),
+	                              std_fs::path(path + SampleExtension),
+	                              get_resource_path(DiskNoiseDir, path),
+	                              get_resource_path(DiskNoiseDir,
+	                                                path + SampleExtension)};
+
+	for (const auto& candidate : candidate_paths) {
+		if (!std_fs::exists(candidate)) {
+			continue;
+		}
+
+		// --- Load entire file into memory ---
+		std::ifstream file(candidate, std::ios::binary | std::ios::ate);
+		if (!file) {
+			LOG_WARNING("DISKNOISE: Failed to open file %s",
+			            candidate.c_str());
+			continue;
+		}
+
+		std::streamsize size = file.tellg();
+		file.seekg(0, std::ios::beg);
+
+		std::vector<uint8_t> file_data(static_cast<size_t>(size));
+		if (!file.read(reinterpret_cast<char*>(file_data.data()), size)) {
+			LOG_WARNING("DISKNOISE: Failed to read file %s",
+			            candidate.c_str());
+			continue;
+		}
+
+		// --- Decode FLAC from memory ---
+		drflac* decoder = drflac_open_memory(file_data.data(),
+		                                     file_data.size(),
+		                                     nullptr);
+		if (!decoder) {
+			LOG_WARNING("DISKNOISE: drflac failed to parse FLAC from %s",
```

Comment:
Just `DISKNOISE: Failed to parse FLAC file '%s'`

We enclose filenames with apostrophes in log statements, so `'%s'`.